### PR TITLE
CORTX-30332: POC for operations quota

### PR DESCRIPTION
**Pull request by avrnv**
# Pull Request
## Problem Statement
-   [CORTX-30332](https://jts.seagate.com/browse/CORTX-30332): number of requests per second.
## Design
-   Add middleware that counts the number of requests currently processed and declines requests if quota is full.

-   \[ \] Coding conventions are followed and code is consistent

-   \[x\] Unit and System Tests are added
-   \[x\] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   \[x\] Testing was performed with RPM

-   \[ \] Interface change (if any) are documented
-   \[ \] Side effects on other features (deployment/upgrade)
-   \[ \] Dependencies on other component(s)

-   \[x\] JIRA number/GitHub Issue added to PR
-   \[x\] PR is self reviewed
-   \[x\] Jira and state/status is updated and JIRA is updated with PR link
-   \[x\] Check if the description is clear and explained

-   \[ \] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/core/blogic/const.py [hunk 1]
```diff
@@ -964,3 +964,9 @@
 
 # CSM usage quotas
 CSM_ACTIVE_USERS_QUOTA = 50
+CSM_ACTIVE_REQUESTS_QUOTA = 100
+MAX_MEM_PER_REQUEST = 5
+
+USAGE_REQUEST_QUOTA = 'usage>request_quota'
```
Comment (Pranali04796) [inline csm/core/blogic/const.py:970]: @avrnv Please use same key name in csm config as that of global configuration. That way it is easy to handle that during upgrade. Covered this change for all other keys in my upgrade PR.

File: csm/conf/configure.py [hunk 2]
```diff
@@ -265,3 +266,55 @@ def create_topics(self):
...
+        limits = Conf.get(const.CONSUMER_INDEX, const.CSM_USAGE_LIMIT_SERVICES)
+        limits = next(filter(lambda x: x[const.NAME] == "agent", limits), None)
+        if limits:
+            mem_min = Configure._mem_limit_to_int(limits["memory"]["min"])
+            mem_max = Configure._mem_limit_to_int(limits["memory"]["max"])
+            cpu_min = Configure._cpu_limit_to_int(limits["cpu"]["min"])
+            cpu_max = Configure._cpu_limit_to_int(limits["cpu"]["max"])
+
+            request_rate = Configure._calculate_request_quota(mem_min, mem_max, cpu_min, cpu_max)
+            Conf.set(const.CSM_GLOBAL_INDEX, const.USAGE_REQUEST_QUOTA, request_rate)
```
Comment (Pranali04796) [inline csm/conf/configure.py:319]: instead of const.USAGE_REQUEST_QUOTA use const.CSM_USAGE_LIMIT_SERVICES.

File: csm/core/agent/api.py [hunk 3]
```diff
@@ -108,8 +111,14 @@ def init():
         CsmRestApi._bgtasks = []
         CsmRestApi._wsclients = WeakSet()
 
+        request_quota = int(Conf.get(const.CSM_GLOBAL_INDEX, const.USAGE_REQUEST_QUOTA))
```
Comment (Pranali04796) [inline csm/core/agent/api.py:114]: instead of const.USAGE_REQUEST_QUOTA use const.CSM_USAGE_LIMIT_SERVICES.

File: csm/conf/configure.py (hunk 2 above)
Comment (avrnv) [reply csm/conf/configure.py:319]: @Pranali04796 , CSM_USAGE_LIMIT_SERVICES is a destination of an array with limits for all CSM services (currently only agent). Global configuration does not include the quota, so there is no entity with the same meaning as USAGE_REQUEST_QUOTA.

File: csm/core/blogic/const.py (hunk 1 above)
Comment (avrnv) [reply csm/core/blogic/const.py:970]: @Pranali04796 , CSM_USAGE_LIMIT_SERVICES is a destination of an array with limits for all CSM services (currently only agent). Global configuration does not include the quota, so there is no entity with the same meaning as USAGE_REQUEST_QUOTA.

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/blogic/const.py [hunk 4]
```diff
@@ -964,3 +964,9 @@
 
 # CSM usage quotas
 CSM_ACTIVE_USERS_QUOTA = 50
+CSM_ACTIVE_REQUESTS_QUOTA = 100
+MAX_MEM_PER_REQUEST = 5
```
[the remark this answers is not in the document]
Comment (avrnv) [reply csm/core/blogic/const.py]: Done.

**Review by Pranali04796 — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**